tools/cmux.c

```diff
@@ -0,0 +1,434 @@
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <errno.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <termios.h>
+#include <unistd.h>
+
+#include <sys/ioctl.h>
+#include <sys/stat.h>
+#include <sys/sysmacros.h>
+
+#include <linux/gsmmux.h>
+#include <linux/tty.h>
+
+/* Maximum transfert unit (MTU), value in bytes */
+
+#define MTU 127
+#define MRU 127
+
+/* Name of the virtual TTYs to create */
+
+#define DEVICE_NAME "/dev/ttyGSM"
+
+/* Name of the driver, used to get the major number */
+
+#define DRIVER_NAME "gsmtty"
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: get_major
+ *
+ * Description:
+ *   Gets the major number of the driver device.
+ *   Returns the major number on success, -1 on failure.
+ *
+ ****************************************************************************/
+
+static int get_major(const char *driver)
+{
+  char device[NAME_MAX];
+  char *line = NULL;
+  int major = -1;
+  size_t len = 0;
+  ssize_t read;
+  FILE *fp;
+
+  /* Open /proc/devices file */
+
+  if ((fp = fopen("/proc/devices", "r")) == NULL)
+    {
+      perror("Cannot open /proc/devices\n");
+      return -1;
+    }
+
+  while ((major == -1) && (read = getline(&line, &len, fp)) != -1)
+    {
+      if (strstr(line, driver) != NULL)
+        {
+          if (sscanf(line, "%d %s\n", &major, device) != 2)
+            {
+              major = -1;
+            }
+        }
+
+      /* Free the line before getting a new one */
+
+      free(line);
+      line = NULL;
+    }
+
+  fclose(fp);
+  return major;
+}
+
+/****************************************************************************
+ * Name: make_nodes
+ *
+ * Description:
+ *   Creates the virtual TTY nodes.
+ *   Returns the number of nodes created.
+ *
+ ****************************************************************************/
+
+int make_nodes(int major, const char *basename, int number)
+{
+  char devname[PATH_MAX];
+  int created = 0;
+  mode_t oldmask;
+  dev_t device;
+  int minor;
+
+  /* Set a new mask to get 666 mode and stores the old one */
+
+  oldmask = umask(0);
+
+  for (minor = 1; minor <= number; minor++)
+    {
+      /* Append the minor number to the base name */
+
+      sprintf(devname, "%s%d", basename, minor);
+
+      /* Store a device info with major and minor */
+
+      device = makedev(major, minor);
+
+      /* Create the actual character node */
+
+      if (mknod(devname, S_IFCHR | 0666, device) != 0)
```

Comment:
@xiaoxiang781216 I think mknod will require root access, doesn't it?

Reply:
Yes, cmux tool requires root privileges to execute

Reply:
@Gary-Hobson thank you for confirming! @xiaoxiang781216 this information also should be included in the Documentation/ ! Please update the patch adding the documentation entry.